genapi/listener_reloader.go

```diff
@@ -58,9 +61,21 @@ func (lr *listenerReloader) Accept() (net.Conn, error) {
 
 // Close only closes the wrapping net.Listener, it's assumed the underlying one
 // will be subsequently closed down the chain
+// In the event of http.Server.Shutdown, Close will be called by the defer in
+// Serve and the Shutdown function. So we detect if we're already closed and
+// return an error if we already are. This seems consistant with builtin.
 func (lr *listenerReloader) Close() error {
-	close(lr.closeCh)
-	return (<-lr.lch).Close()
+	// If we are already closed then there won't be anything to read on lch,
```

Comment:
There's still _kind of_ a race condition with this I think:

* Process A calls Close(), reads a listener off lch
* Process B calls Accept(), reads a listener off lch, calls Accept() on the listener
* Process A (still in Close) calls Close() on that listener
* something weird happens, cause I dunno if you can call Accept() and Close() on a listener at the same time.

What we could do is have closeCh be a `chan net.Listener` which spin is always trying to write to, and once it's written to it spin assumes it should return out.